### PR TITLE
Systemd testsuite v244 pull

### PR DESCRIPTION
**Pull request by tblume**
systemd testsuite for openSUSE Tumbleweed

This is for systemd version 244 that will soon reach tumbleweed.
The current version 243 in tumbleweed wasn't fully tested with this code.

- Related ticket: https://progress.opensuse.org/issues/45158

- Needles: none

- Verification run:

x86_64
http://emerson.suse.de/tests/2010

aarch64
http://emerson.suse.de/tests/2015

The testcode uses some custom Variables, please see command line below (aarch64 test):

su geekotest -c "/usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org/tests 1138005 SYSTEMD_TESTSUITE=noqemu QA_TESTSUITE_REPO=https://download.opensuse.org/repositories/home:/tsaupe:/branches:/devel:/openSUSE:/QA:/Tumbleweed/openSUSE_Factory_ARM NUMDISKS=2 HDD_1=systemd-testsuite-second-disk.raw MAX_JOB_TIME=14400 SYSTEMD_FROM_TESTREPO=1"

Explanation:

 SYSTEMD_TESTSUITE=noqemu - run without nested qemu
QA_TESTSUITE_REPO=  - my custom development repo for the testsuite package 
(this Variable can be remove as soon as https://build.opensuse.org/request/show/765156 was accepted)
NUMDISKS=2 - cryptsetup test need a second disk image
HDD_1=systemd-testsuite-second-disk.raw - second disk image for cryptsetup test
MAX_JOB_TIME=14400 - this test easily exceeds the default 2 hours test run on slower machines
SYSTEMD_FROM_TESTREPO=1" - install the systemd package that was built together with the           systemd-qa-package (normally not needed in official openQA runs)

### CODE REVIEW

File: lib/bootloader_setup.pm [hunk 1]
```diff
@@ -916,30 +916,33 @@ sub tianocore_http_boot {
...
-            $svirt->add_disk({file => $patched_img, dev_id => 'a'});
+    my $dev_id = 'a';
+    my $numdisks = get_var('NUMDISKS') || '1';
+
+    foreach my $n (1 .. $numdisks) {
+        if (my $hdd = get_var('HDD_' . $n)) {
+            my $basename = basename($hdd);
+            my $basedir  = svirt_host_basedir();
+            my $hdd_dir  = "$basedir/openqa/share/factory/hdd";
+            my $hdd_dir  = "/var/lib/libvirt/images";
```
Comment (foursixnine) [inline lib/bootloader_setup.pm]: You're using the same variable twice, or rather you're trying to change the behavior here?

File: tests/systemd_testsuite/build_from_upstream_git.pm [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+use known_bugs;
+use testapi;
+use power_action_utils 'power_action';
+use utils 'zypper_call';
+use version_utils qw(is_opensuse is_sle is_tumbleweed);
+
+
+sub installandbuild {
+    zypper_call 'in osc';
+    assert_script_run "osc branch openSUSE:Factory systemd";
```
Comment (foursixnine) [inline tests/systemd_testsuite/build_from_upstream_git.pm]: you want to build locally, I hope?.

File: lib/bootloader_setup.pm (hunk 1 above)
Comment (tblume) [reply lib/bootloader_setup.pm]: Oops, seems that was a remainder of obsolete testcode. I've removed it now.

File: tests/systemd_testsuite/build_from_upstream_git.pm (hunk 2 above)
Comment (tblume) [reply tests/systemd_testsuite/build_from_upstream_git.pm]: Actually that is a preparation for my planned testservice for systemd upstream, which will be on a dedicated local machine.
It isn't used in the tumbleweed test.
Should I remove build_from_upstream_git.pm  from this submission or can I leave it to continue this later on?

Comment (tblume) [reply tests/systemd_testsuite/build_from_upstream_git.pm]: Ok, I've removed the build_from_upstream_git.pm file. It doesn't belong to the tumbleweed tests and only causes confusion.

Comment (foursixnine) [reply tests/systemd_testsuite/build_from_upstream_git.pm]: I think if it's not scheduled, there's a script that will complain about the test not being scheduled :P

**Comment by tblume** [comment]
Verification run from offical openQA repo using command line:
su geekotest -c "/usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org/tests 1150891 SYSTEMD_TESTSUITE=noqemu NUMDISKS=2 HDD_2=systemd-testsuite-second-disk.raw MAX_JOB_TIME=14400"

http://emerson.suse.de/tests/2024

**Comment by SergioAtSUSE** [comment]
I somehow missed you last comment.

LGTM, but you need to fix the whitespaces that travis mentions.
https://travis-ci.org/os-autoinst/os-autoinst-distri-opensuse/jobs/639501055#L464

**Comment by tblume** [comment]
> I somehow missed you last comment.
> 
> LGTM, but you need to fix the whitespaces that travis mentions.
> https://travis-ci.org/os-autoinst/os-autoinst-distri-opensuse/jobs/639501055#L464

Ok, I've updated my pr and travis seems to be satisfied now.

**Comment by dzedro** [comment]
https://openqa.suse.de/tests/3871825#step/binary_tests/1

**Comment by tblume** [comment]
> https://openqa.suse.de/tests/3871825#step/binary_tests/1

Please note that this pr is for testing systemd version 244 in factory/tumbleweed. SLES15 has the way older systemd version 234. I didn't try my testcode with this version.

**Comment by dzedro** [comment]
@tblume  I think it didn't even come to testing :)

**Comment by SergioAtSUSE** [comment]
@tblume, that happened also for SLE12, when the tests were not adapted for version under maintenance.

I try to inform people that these tests are only for integration testing of products under development, since this test suite is just an adapted for openQA version of upstream tests.

Do you want to maintain them also for released products?

**Comment by tblume** [comment]
> @tblume, that happened also for SLE12, when the tests were not adapted for version under maintenance.
> 
> I try to inform people that these tests are only for integration testing of products under development, since this test suite is just an adapted for openQA version of upstream tests.
> 
> Do you want to maintain them also for released products?

Hm, actually it would be nice to have it all in one code, but I'm not sure yet how much the effort would be. SLES15 is 10 versions down from upstream and SLES12 even more.
I will test whether applying the upstream testsuite is feasible.

**Comment by vpelcak** [comment]
Hello.

Is there any progress with this?
This PR broke tests for QAM.

As agreed on:
https://confluence.suse.com/display/openqa/Testcase+Maintenance+-+Policy

PR is supposed to be retracted by its creator and fixed with help on verification with the one who reported the problem.

Could you please work on that?

**Comment by SergioAtSUSE** [comment]
> Hello.
> 
> Is there any progress with this?
> This PR broke tests for QAM.
> 
> As agreed on:
> https://confluence.suse.com/display/openqa/Testcase+Maintenance+-+Policy
> 
> PR is supposed to be retracted by its creator and fixed with help on verification with the one who reported the problem.
> 
> Could you please work on that?

@tblume ?

**Comment by tblume** [comment]
Maybe we have a misunderstanding here:

https://openqa.suse.de/tests/3871825#step/binary_tests/1

is invalid for this testcode, because SLES15 is using systemd version 234 and not 244.
This test is valid for systemd version 244 in openSUSE tumbleweed.
It is not intended to be used in SLES15.

Or do you mean another test issue?
If so, where was it reported?

**Comment by vpelcak** [comment]
> Maybe we have a misunderstanding here:
> 
> https://openqa.suse.de/tests/3871825#step/binary_tests/1
> 
> is invalid for this testcode, because SLES15 is using systemd version 234 and not 244.
> This test is valid for systemd version 244 in openSUSE tumbleweed.
> It is not intended to be used in SLES15.
> 
> Or do you mean another test issue?
> If so, where was it reported?

Please help me to understand.

The way I understand it:
1) There was a systemd testsuite working for SLE
2) This PR updated it to work on Tumbleweed
3) Now the test no longer works for SLE

Do I understand it right?

**Comment by tblume** [comment]
Hm, I thought you have a separated git branch for SLE.
So, the same openQA code is used for openSUSE and SLE?
If so, the issue should be already fixed by:

https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/9822

**Comment by okurz** [comment]
yes, given that maintenance updates are also introducing the need for test updates it is more maintainable to use the same git branch but only introduce logical branching with if/else where needed.

**Comment by tblume** [comment]
> yes, given that maintenance updates are also introducing the need for test updates it is more maintainable to use the same git branch but only introduce logical branching with if/else where needed.

Ok thanks, so I need to take care about backward compatibility next time.

**Pull request merged**